### PR TITLE
uuid.cpp: GenerateRandomUUID: fix indexing

### PR DESCRIPTION
**Pull request by nodakai**
DuckDB console printed wrongly formatted UUID v4.1 values
```
D SELECT uuid() FROM range(10);
┌──────────────────────────────────────┐
│                uuid()                │
├──────────────────────────────────────┤
│ a4b889d1-6144-d390-904b-a3fdef530bdd │
│ ea5d7657-094e-17ba-bafd-6ef3b3b3a81b │
│ 3b230544-5946-da9b-9be5-30a0efdcd79e │
│ 463e1758-7349-93bb-bb85-872365d20f3f │
│ e9fba125-f846-0fa6-a67d-cd29fe5b4c9a │
│ 63759849-f444-b7a4-a432-76cb34bc2f3a │
│ f6034e52-0d40-a79b-9b1a-cc0ec14be746 │
│ 3c89c15d-9c44-51b3-b3ec-4c74a7561269 │
│ 1d0cebe4-834e-c48d-8ddd-a14b83c07a5e │
│ 291aa393-e046-31ba-bab8-14453aa6e05e │
│ c1352366-5143-83bd-bd39-8cdde39a55b4 │
```

According to [Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)), UUID v4.1 values should look like
`XXXXXXXX-XXXX-4XXX-[89ab]XXX-XXXXXXXXXXXX` but the pattern observed in the above transcript is
`XXXXXXXX-XXXX-XX4X-XX[89ab]X-XXXXXXXXXXXX`

Python stdlib agreed with Wikipedia's description:

```
>>> for _ in range(10):
...   print(uuid.uuid4())
... 
b6881b6f-5c96-49f5-a421-b749aa31be38
bec73b38-ec4f-4411-a7a5-e89f08cbab93
0a1e42be-fc90-49d9-9811-648ce568d5fe
5d97f674-fe36-4cc5-a2f8-e1270dcf519c
266dd212-c7de-4f8f-b16f-099c67c1e76b
dbbd4398-e4cc-4a90-802b-e729d65243e2
cde64de4-4328-4e73-8d78-aecede4c92c9
01e65e02-d272-4228-95c9-07ce7ad7288b
942b2f6f-e1ae-4a36-a5ed-c709a7b86d53
ebeab62a-1033-477f-b6d5-70a107f4fc61
```

@Mytherin noted Python binding generated Python UUID objects which, when printed on Python console, exhibited the same formatting errors, and alleged the issue was in UUID v4 generation itself rather than printing.

```
>>> import duckdb
>>> con = duckdb.connect()
>>> con.execute('SELECT uuid() FROM range(10)').df()
                                 uuid()
0  97ae45f9-6148-6abd-bd6e-9129c72bd9fa
1  3cd5d43c-6145-119d-9daf-749959a7eb31
2  315f527f-a345-448e-8e8f-7b19ef1f2a07
3  15a39ffb-ef40-4d88-8819-23348439e9d3
4  e45c3255-5641-0e9f-9f4e-c2a76e44cc50
5  3affc0e7-544c-8e95-9555-b5e9a45511d4
6  81942ff6-2e46-f19b-9b74-6a136bff4a08
7  e1a6e5cd-b04f-b9b9-b9d9-efd8fd39d2b8
8  049582b4-4b4b-2ebf-bf09-4b429e519377
```

Then I could find and fix the indexing errors in it.

https://discord.com/channels/909674491309850675/921073327009853451/1027246112416997488

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

Could we also add a test that verifies this works as intended? For example, we can look at the string converted to varchar and verify the property holds as described. The following query should always return "4", for example:

```sql
select min(substring(gen_random_uuid()::varchar, 15, 1)) from generate_series(0,100,1);
-- 4
```

**Comment by nodakai** [comment]
Added tests to `test/sql/function/uuid/test_uuid.test`; let me know if you think there's a better place for them

File: test/sql/function/uuid/test_uuid.test [hunk 1]
```diff
@@ -42,3 +42,19 @@ SELECT * FROM uuids ORDER BY gen_random_uuid()
...
+# UUID v4 follows `XXXXXXXX-XXXX-4XXX-XXXX-XXXXXXXXXXXX` format
+query I
+SELECT DISTINCT substring(uuid()::varchar, 15, 1) FROM range(100)
+----
+4
+
+# UUID v4 variant 1 follows `XXXXXXXX-XXXX-4XXX-[89ab]XXX-XXXXXXXXXXXX` format
+# the probability for 100 iters not to cover the 4 possible values is (3/4)**100 == 3e-13
+query I
+SELECT DISTINCT substring(uuid()::varchar, 20, 1) AS x FROM range(100) ORDER BY x
```
[the remark this answers is not in the document]
Comment (nodakai) [reply test/sql/function/uuid/test_uuid.test:55]: The probability for 100 iterations not to yield 4 distinct values is in the order of (3/4)**100 == 3e-13 if my math didn't fail me, and if I were you I'd want to be notified of such a failure (I'd see it as a sign that rng isn't working as expected).
If my comment didn't convey my intention well, please suggest how I can improve it

[the remark this answers is not in the document]
Comment (Mytherin) [reply test/sql/function/uuid/test_uuid.test:55]: The point about detecting if the RNG is not working as expected is indeed fair. Once every trillion iterations is probably rare enough :)

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**